vision_agent/tools/meta_tools.py

```diff
@@ -0,0 +1,402 @@
+import os
+import subprocess
+from pathlib import Path
+from typing import Any, Dict, List, Union
+
+import vision_agent as va
+from vision_agent.lmm.types import Message
+from vision_agent.tools.tool_utils import get_tool_documentation
+from vision_agent.tools.tools import TOOL_DESCRIPTIONS
+
+# These tools are adapted from SWE-Agent https://github.com/princeton-nlp/SWE-agent
+
+CURRENT_FILE = None
+CURRENT_LINE = 0
+DEFAULT_WINDOW_SIZE = 100
+ZMQ_PORT = os.environ.get("ZMQ_PORT", None)
+
+
+def report_progress_callback(port: int, inp: Dict[str, Any]) -> None:
+    import zmq
+
+    context = zmq.Context()
+    socket = context.socket(zmq.PUSH)
+    socket.connect(f"tcp://localhost:{port}")
+    socket.send_json(inp)
+
+
+def filter_file(file_name: Union[str, Path]) -> bool:
+    file_name_p = Path(file_name)
+    return (
+        file_name_p.is_file()
+        and "__pycache__" not in str(file_name_p)
+        and file_name_p.suffix in [".py", ".txt"]
+        and not file_name_p.name.startswith(".")
+    )
+
+
+def generate_vision_code(save_file: str, chat: str, media: List[str]) -> str:
+    """Generates python code to solve vision based tasks.
+
+    Parameters:
+        save_file (str): The file path to save the code.
+        chat (str): The chat message from the user.
+        media (List[str]): The media files to use.
+
+    Returns:
+        str: The generated code.
+
+    Examples
+    --------
+        >>> generate_vision_code("code.py", "Can you detect the dogs in this image?", ["image.jpg"])
+        from vision_agent.tools import load_image, owl_v2
+        def detect_dogs(image_path: str):
+            image = load_image(image_path)
+            dogs = owl_v2("dog", image)
+            return dogs
+    """
+
+    if ZMQ_PORT is not None:
```

Comment:
Are you using ZMQ to keep the underlying call asynchronous as we will not know when the vision agent coder will respond ?

Reply:
Yes, so when VisionAgent calls VisionAgentCoder we don't get any response from it because it runs in another process, so it's silent for a long time. Here we can use the Code Execution Logs to monitor the progress by reading from the zmq context